File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
def dfs(a,b):
    if a<0 or b<0 or a>=n or b>=m:
        return False
    if arr[a][b] == 1:
        arr[a][b] = 0 

        dfs(a-1,b)
        dfs(a+1,b)
        dfs(a,b+1)
        dfs(a,b-1)

        return True
    return False

t = int(input())
answer = []
for i0 in range(t):
    n,m,k = map(int,input().split())
    arr = [[0 for _ in range(50)] for __ in range(50)]

    for i1 in range(k):
        x,y = map(int,input().split())
        arr[x][y] = 1

    result = 0
    for i in range(n):
        for j in range(m):
            if dfs(i,j):
                result += 1
    answer.append(result)

for i in answer:
    print(i)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.util.Scanner;

public class Main {
    static int n, m;
    static int[][] arr;
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        int[] answer = new int[t];
        
        for (int i0 = 0; i0 < t; i0++) {
            n = scanner.nextInt();
            m = scanner.nextInt();
            int k = scanner.nextInt();
            
            arr = new int[50][50];
            
            for (int i1 = 0; i1 < k; i1++) {
                int x = scanner.nextInt();
                int y = scanner.nextInt();
                arr[x][y] = 1;
            }
            
            int result = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (dfs(i, j)) {
                        result += 1;
                    }
                }
            }
            answer[i0] = result;
        }
        
        for (int i : answer) {
            System.out.println(i);
        }
    }

    public static boolean dfs(int a, int b) {
        if (a < 0 || b < 0 || a >= n || b >= m) {
            return false;
        }
        if (arr[a][b] == 1) {
            arr[a][b] = 0;
            
            dfs(a - 1, b);
            dfs(a + 1, b);
            dfs(a, b + 1);
            dfs(a, b - 1);
            
            return true;
        }
        return false;
    }
}